common/metagraph_syncer.py

```diff
@@ -0,0 +1,145 @@
+
+        Args:
+            subtensor (bt.subtensor): The subtensor used to fetch the metagraphs.
+            config (Dict[int, int]): A mapping of netuid to the cadence (in seconds) to sync the metagraph.
+        """
+        self.subtensor = subtensor
+        self.config = config
+        self.metagraph_map: Dict[int, MetagraphSyncer._State] = {
+            netuid: MetagraphSyncer._State() for netuid in config.keys()
+        }
+        self.is_running = False
+        self.done_initial_sync = False
+        self.lock = threading.RLock()
+        self.notify_listener_thrad_pool = ThreadPoolExecutor(max_workers=1)
+
+    def do_initial_sync(self):
+        """Performs an initial sync of all metagraphs.
+
+        Unlike regular syncs, this will not notify listeners of the updated metagraph.
+        """
+        bt.logging.debug("Metagraph syncer do_initial_sync called")
+
+        for netuid in self.config.keys():
+            fn = functools.partial(self.subtensor.metagraph, netuid)
+            with self.lock:
+                metagraph = utils.run_in_subprocess(fn, ttl=120)
+                state = self.metagraph_map[netuid]
+                state.metagraph = metagraph
+                state.last_synced_time = datetime.now()
+
+            bt.logging.debug(f"Successfully loaded metagraph for {netuid}")
+
+        self.done_initial_sync = True
+
+    def start(self):
+        bt.logging.debug("Metagraph syncer start called")
+
+        assert self.done_initial_sync, "Must call do_initial_sync before starting"
+
+        self.is_running = True
+        thread = threading.Thread(target=self._run, daemon=True)
+        thread.start()
+
+    async def _sync_metagraph_loop(self, netuid: int, cadence: int):
+        while self.is_running:
+            # On start, wait cadence before the first sync.
+            await asyncio.sleep(cadence)
+
+            try:
+                # Intentionally block the shared thread so that we only
+                # sync 1 metagraph at a time.
+                metagraph = utils.run_in_subprocess(
+                    functools.partial(self.subtensor.metagraph, netuid), ttl=120
+                )
+                state = None
+                with self.lock:
+                    # Store metagraph and sync time
+                    state = self.metagraph_map[netuid]
+                    state.metagraph = metagraph
+                    state.last_synced_time = datetime.now()
+
+                await self._notify_listeners(state, netuid)
+
+                await asyncio.sleep(cadence)
+            except (BaseException, Exception) as e:
+                bt.logging.error(
+                    f"Error when syncing metagraph for {netuid}: {e}. Retrying in 60 seconds."
+                )
+                await asyncio.sleep(60)
+
+    def _run(self):
+        # For each netuid we should sync metagraphs for, spawn a Task to sync it.
+        asyncio.gather(
```

Comment:
Should this just be wait if we aren't using the gathered return?

Reply:
Good catch - didn't realize there was a wait() API.